File: FundComponents/ImpactExtratropicalStormsComponent.cs
// FUND - Climate Framework for Uncertainty, Negotiation and Distribution
// Copyright (C) 2012 David Anthoff and Richard S.J. Tol
// http://www.fund-model.org
// Licensed under the MIT license
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using Esmf;
using Fund.CommonDimensions;

namespace Fund.Components.ImpactExtratropicalStorms
{

    public interface IImpactExtratropicalStormsState
    {
        IVariable2Dimensional<Timestep, Region, double> extratropicalstormsdam { get; }
        IVariable2Dimensional<Timestep, Region, double> extratropicalstormsdead { get; }

        IParameter1Dimensional<Region, double> extratropicalstormsbasedam { get; }
        Double extratropicalstormsdamel { get; }
        IParameter1Dimensional<Region, double> extratropicalstormspar { get; }
        IParameter1Dimensional<Region, double> extratropicalstormsbasedead { get; }
        Double extratropicalstormsdeadel { get; }
        Double extratropicalstormsnl { get; }

        IParameter1Dimensional<Region, double> gdp90 { get; }
        IParameter1Dimensional<Region, double> pop90 { get; }

        IParameter2Dimensional<Timestep, Region, double> population { get; }
        IParameter2Dimensional<Timestep, Region, double> income { get; }

        IParameter1Dimensional<Timestep, double> acco2 { get; }
        Double co2pre { get; }
    }


    public class ImpactExtratropicalStormsComponent
    {
        public void Run(Clock clock, IImpactExtratropicalStormsState state, IDimensions dimensions)
        {
            var t = clock.Current;
            var s = state;

            foreach (var r in dimensions.GetValues<Region>())
            {
                double ypc = s.income[t, r] / s.population[t, r] * 1000.0;
                double ypc90 = s.gdp90[r] / s.pop90[r] * 1000.0;

                s.extratropicalstormsdam[t, r] = s.extratropicalstormsbasedam[r] * s.income[t, r] * Math.Pow(ypc / ypc90, s.extratropicalstormsdamel) * (Math.Pow(1.0 + (s.extratropicalstormspar[r] * (s.acco2[t] / s.co2pre)), s.extratropicalstormsnl) - 1.0);
                s.extratropicalstormsdead[t, r] = 1000.0 * s.extratropicalstormsbasedead[r] * s.population[t, r] * Math.Pow(ypc / ypc90, s.extratropicalstormsdeadel) * (Math.Pow(1.0 + (s.extratropicalstormspar[r] * (s.acco2[t] / s.co2pre)), s.extratropicalstormsnl) - 1.0);
            }
        }
    }
}

File: src/components/ImpactExtratropicalStormsComponent.jl
@defcomp impactextratropicalstorms begin
    regions = Index()

    extratropicalstormsdam = Variable(index=[time,regions])
    extratropicalstormsdead = Variable(index=[time,regions])

    extratropicalstormsbasedam = Parameter(index=[regions])
    extratropicalstormsdamel = Parameter(default=-0.514)
    extratropicalstormspar = Parameter(index=[regions])
    extratropicalstormsbasedead = Parameter(index=[regions])
    extratropicalstormsdeadel = Parameter(default=-0.501)
    extratropicalstormsnl = Parameter(default=1)

    gdp90 = Parameter(index=[regions])
    pop90 = Parameter(index=[regions])

    population = Parameter(index=[time,regions])
    income = Parameter(index=[time,regions])

    acco2 = Parameter(index=[time])
    co2pre = Parameter()

    function run_timestep(p, v, d, t)

        for r in d.regions
            ypc = p.income[t, r] / p.population[t, r] * 1000.0
            ypc90 = p.gdp90[r] / p.pop90[r] * 1000.0

            v.extratropicalstormsdam[t, r] = p.extratropicalstormsbasedam[r] * p.income[t, r] * (ypc / ypc90)^p.extratropicalstormsdamel * ((1.0 + (p.extratropicalstormspar[r] * (p.acco2[t] / p.co2pre)))^p.extratropicalstormsnl - 1.0)
            v.extratropicalstormsdead[t, r] = 1000.0 * p.extratropicalstormsbasedead[r] * p.population[t, r] * (ypc / ypc90)^p.extratropicalstormsdeadel * ((1.0 + (p.extratropicalstormspar[r] * (p.acco2[t] / p.co2pre)))^p.extratropicalstormsnl - 1.0)
        end
    end
end
